src/Storages/PartitionCommands.h

```diff
@@ -66,9 +67,39 @@ struct PartitionCommand
     String move_destination_name;
 
     static std::optional<PartitionCommand> parse(const ASTAlterCommand * command);
+    /// Convert type of the command to string (use not only type, but also
+    /// different flags)
+    std::string typeToString() const;
 };
 
 using PartitionCommands = std::vector<PartitionCommand>;
 
+/// Result of exectuin of a single partition commands. Partition commands quite
+/// different, so some fields will be empty for some commands. Currently used in
+/// ATTACH and FREEZE commands.
+struct PartitionCommandResultInfo
+{
+    /// Command type, always filled
+    String command_type;
+    /// Partition id, always filled
+    String partition_id;
+    /// Part name, always filled
+    String part_name;
+    /// Part name in /detached directory, filled in ATTACH
+    String old_part_name;
+    /// Path to backup directory, filled in FREEZE
+    String backup_path;
+    /// Name of the backup (specified by user or increment value), filled in
+    /// FREEZE
+    String backup_name;
+};
+
+using PartitionCommandsResultInfo = std::vector<PartitionCommandResultInfo>;
+
+/// Convert partition comands result to Source from single Chunk, which will be
+/// used to print info to the user. Tries to create narrowest table for given
+/// results. For example, if all commands were FREEZE commands, than
+/// old_part_name column will be absent.
+std::shared_ptr<SourceFromSingleChunk> convertCommandsResultToSource(const PartitionCommandsResultInfo & commands_result);
```

Comment:
Probably, we can already return `Pipes` here. No need to specify concrete source type.
(If you really want, declare `SourceFromSingleChunk` and remove `#include <Processors/Sources/SourceFromSingleChunk.h>` from header)